code/game/objects/structures/barsign.dm

```diff
@@ -41,363 +53,717 @@
 
 	update_icon()
 
-/obj/structure/sign/barsign/update_icon_state()
-	. = ..()
-
-	if(!current_sign || broken)
-		icon_state = "empty"
+/obj/machinery/barsign/proc/set_random_sign()
+	if(!emagged)
+		set_sign(pick(barsigns))
+	else
+		set_sign(pick(syndisigns))
+
+// Saves a /datum/barsign to the prev_sign variable.
+/obj/machinery/barsign/proc/save_sign(datum/barsign/S)
+	// Broken, blank, and EMPed screens shouldn't be saved.
+	if(istype(S, /datum/barsign) && !istype(S, /datum/barsign/hiddensigns))
+		prev_sign = S
+		return TRUE
+	return FALSE
+
+/obj/machinery/barsign/update_icon_state()
+	if(build_stage == BARSIGN_FRAME && !istype(current_sign, /datum/barsign/hiddensigns/building/frame))
+		set_sign(new /datum/barsign/hiddensigns/building/frame)
+	if(build_stage == BARSIGN_CIRCUIT && !istype(current_sign, /datum/barsign/hiddensigns/building/circuited))
+		set_sign(new /datum/barsign/hiddensigns/building/circuited)
+	if(build_stage == BARSIGN_WIRED && !istype(current_sign, /datum/barsign/hiddensigns/building/wired))
+		set_sign(new /datum/barsign/hiddensigns/building/wired)
```

Comment:
You intended for these three to return too? If it is one then it for sure wont be the other and set_sign willl trigger an update_icon that will make it reach the "icon_state = current_sign.icon"